dips/dip-16.md

```diff
@@ -0,0 +1,112 @@
+    * E.g., `f(u64, signer)`, `f(signer, u64, signer)`, `f(signer, bool, u64, signer)` do not satisfy this condition
+* Each non-`signer` type in `param_types` has the `copy` ability and is not a struct
+    * E.g., `f(signer, Diem<XUS>`) does not satisfy this condition because `Diem<XUS>` is a struct
+* Each type in `param_types` is *closed* (that is, does not refer to the type variables in `ability_params`
+    * E.g., `f<T>(signer, vector<T>)` does not satisfy this condition
+    * E.g., `f(signer, vector<u8>`) does satisfy this condition
+
+## Script Functions
+
+In the Move language, a function can be declared with four different visibility levels: `public`, `private`, `public(friend)`, and `public(script)`. *Any* function with `public(script)` visibility can be invoked by a transaction if it satisfies the script signature requirements explained below.
+
+A client can define a `ScriptFunction` transaction payload using the following struct:
+
+```
+pub struct ScriptFunction {
+    module: ModuleId,
+    function: Identifier,
+    ty_args: Vec<TypeTag>,
+    args: Vec<Vec<u8>>,
+}
+```
+
+Here, a `module` is an (account address, module name) pair that refers to a module already published on chain. The `function` should be the name of a function declared by `module` with script visibility. The signature of `function` must satisfy the script signature requirements described below.
+
+The type arguments `ty_args` are a list of Move types. Each type must either be a valid Move ground type (e.g., `u64`) or the fully qualified identifier of a type declared on-chain (e.g., `0x1::DiemAccount::DiemAccount`). In addition, the length of `ty_args` and `ability_params` must be the same and each type must satisfy the constraints expressed by `ability_params`.
+
+The value arguments `args` are a list of BCS-encoded values. The length of `args` and `param_types` must be the same. Each value argument must deserialize correctly according to its corresponding type in `param_types`.
+
+## Transaction Scripts
+
+A transaction script is a Move bytecode program that contains a single function `f`. The signature of `f` must satisfy the script signature requirements described below.
+
+A client can define a `Script` transaction payload using the following struct.
+
+```
+pub struct Script {
+    code: Vec<u8>,
+    ty_args: Vec<TypeTag>,
+    args: Vec<TransactionArgument>,
+}
+```
+
+Here, `code` is a Move bytecode program. The `ty_args` are the same as for script functions. The `args` are a list of `TransactionArgument`s instead of a list of BCS-encoded values.
+
+Only a restricted set of programs can be included in `code`. The allowed programs are specified in an on-chain allowlist published at at `0x1::DiemTransactionPublishingOption::DiemTransactionPublishingOption.script_allowlist`. The allowlist contains the sha3-256 hash of each program rather than the code itself.
```

Comment:
I just noticed the repeated "at at" here.

More importantly, should we mention that DiemRoot is exempted from the allowlist check?

Reply:
Very good point!